man/kunoichi.1.md
Kunoichi(1) -- A process privilege escalation monitor
=====================================================

## SYNOPSIS

`kunoichi` [<OPTION>]...

## DESCRIPTION

Kunoichi is a Ruby rewrite of *ninja(1)*, a privilege escalation monitor.

It is an almost drop-in replacement for Ninja, as in it supports all
the features supported by Ninja and is able to generate its configuration
starting from a Ninja configuration file.

## RUNNING

On startup Kunoichi will look for the configuration file (`config.yml`) in the current working directory and in /etc/kunoichi.

A different configuration file can be supplied with `-c` (see below).

You can start Kunoichi with `-h|--help` to see the runtime options, or with no arguments to run it normally.

## OPTIONS

Kunoichi supports some runtime options to change its behavior, overriding the configuration.

### Configuration overrides:

  * `-n`,`--dry-run`:
    Perform no actions. This sets *no_kill* and *no_kill_ppid* to true.
    
  * `-d`,`--debug`:
    Run in debug mode. This provides extended logging and sets *daemon*, *syslog* and *logfile* to false.

  * `-p`,`--pid-file` <file>:
    Write the pid to an alternate pid file than the one written in the configuration.

  * `-w`,`--whitelist` <file>:
    Use <file> as the whitelist instead of the configured one.

### Miscellaneous options:

  * `-h`,`--help`:
    Show a basic usage help.

  * `-c`,`--config-file` <file>:
    Load a different configuration file than the defaults (see *RUNNING* above).

  * `-i`,`--import` <file>:
    Load a configuration file for *ninja(1)* and print the Kunoichi equivalent configuration on STDOUT.

  * `-g`,`--generate-config` <file>:
    Generates a Kunoichi configuration file. If file is not specified, it writes to *config.yml*.

## FILES

Kunoichi works with two files, *config.yml* and the *whitelist*.

The configuration file is in YAML and can be generated from defaults or from an old ninja configuration (see above).

The whitelist is a list of executables that can be ran by anyone without issuing alerts.

A common example would be:

    # Enable sudo
    /usr/bin/sudo
    # Enable su
    /bin/su


## CONFIGURATION OPTIONS

These are the configuration options understood by Kunoichi:

* **group** - Sets the GID (Group ID) of the group of users that are allowed to run any root setuid/setgid
               binary. Programs executed by users in this group do not need to be whitelisted. _(default: 0)_
* **daemon** - If set to true, Kunoichi will run in the background as daemon. _(default: true)_
* **syslog** - If set to true, Kunoichi will log to syslog using facility _daemon_. _(default: true)_
* **interval** - The process scanning interval, in seconds. Can be float. This is the amount of time Kunoichi will sleep,
                  before going through all the active processes on the system. Lower intervals are safer, meaning
                  that the unauthorized processes will be detected faster. Unless your system is very slow, you should set
                  this to a value between zero and one. _(default: 1)_
* **logfile** - Path to log file. If specified, all messages are logged to this file. Can be used as an alternative to syslog.
                _(default: /var/log/kunoichi.log)_
* **pidfile** - Path to the pid file.  _(default: /var/run/kunoichi.pid)_
* **whitelist** - Path to the whitelist file. _(default: /etc/kunoichi/whitelist.txt)_
* **external_command** - Path to the external command. If you specify an external command, 
                          this command will be run every time an unauthorized process is detected.
                          This could be a script that alerts the admin of the system via e-mail, 
                          disables the login account of the offending user, among other things.
                          The command will have some information on the process available in the environment (see below).
                          _(default: /bin/true)_
* **no_kill** - If this option is enabled, unauthorized processes will not be killed when detected. This can be very
                 useful when testing the system, or when creating whitelists. _(default: true)_
* **no_kill_ppid** - Same as the *no_kill* option, except this one affects the ppid, parent of the unauthorized
                      process. _(default: true)_
* **ignore_root_procs** - When this option is enabled, all processes spawned by root, meaning processes with a parent
                           process owned by root, will be ignored. This is generally a good idea, and will save us some
                           cpu cycles. _(default: true)_
* **log_whitelist** - If enabled, Kunoichi will log every time it detects a whitelisted process running. _(default: true)_
* **require_init_wlist** - This option is very important for the security of your system. When this option is set to *true*,
                            all processes running as root, and controlled by init(1) will have to be whitelisted. If this
                            option is not enabled, attackers can circumvent Kunoichi by forking a new process after gaining
                            root, and create a new session (e.g. using setsid()). This will cause Kunoichi to believe that
                            root spawned this proces, as init(1) will now be the parent, and init(1) is owned by root.
                            Therefore, to gain a higher level of security, you should enable this option and explicitly
                            whitelist every process that will run in its own session, meaning that it will be controlled
                            by init(1). _(default: false)_
* **proc_scan_offset** - This option specifies the offset of which pids to scan. If set to 0, all processes on the system 
                          will be scanned. This might not be necessary, as processes with a pid lower than 300 is generally
                          kernel processes, and we really don't want to keep an eye on those. So, if you want to save a few
                          cpu cycles, set this to 300 or so. _(default: 0)_

Many of the descriptions here are based on ninja documentation, as their behavior is very similar.

## ENVIRONMENT

When an external command is launched, it has these environment variables available:

* *EVIL_PID* - The pid of the offending process
* *EVIL_CMD* - The content of /proc/pid/cmdline, human readable
* *EVIL_NAME* - First entry in cmdline
* *EVIL_BIN* - /proc/pid/exe target of the process
* *EVIL_PPID* - The parent pid of the offending process
* *EVIL_PCMD* - The content of parent's /proc/pid/cmdline, human readable
* *EVIL_PNAME* - First entry in parent's cmdline
* *EVIL_PBIN* - /proc/pid/exe target of the parent process

## COPYRIGHT

Kunoichi is Copyright (C) 2013 Alessandro Grassi <alessandro@aggro.it>.

## LICENSE

Kunoichi is released under the terms of GNU General Public License version 3



lib/kunoichi/kunoichi.rb
module Kunoichi
  # The main class, where all the dirty work happens
  class Daemon

    # Basic validator for configuration
    def validate_conf(config)

      # Base check, might be redundant
      raise 'invalid configuration passed: not an hash' unless config.is_a? Hash
      
      # Check for group validity
      raise 'invalid group: not a number' unless config['group'].is_a? Fixnum
      # Make sure the group exists on /etc/group
      begin
        Etc::getgrgid config['group']
      rescue => e
        raise "invalid group: #{e.message}"
      end

      # Check for daemon validity
      raise 'invalid daemon setting: please set to true or false' unless config['daemon'].is_bool?
      
      # Check interval
      raise 'invalid interval: please set to a number' unless config['interval'].is_a? Fixnum or config['interval'].is_a? Float
      # Useful warning in case no interval was declared
      puts 'WARNING: an interval of 0 can severely impact your machine performance. Use with caution.' if config['interval'] == 0

      # Check syslog
      raise 'invalid value for syslog: please set to true or false' unless config['syslog'].is_bool?

      # Check logfile
      raise 'set either syslog or logfile if daemon mode is enabled' unless config['syslog'] or config['logfile'].is_a? String or !config['daemon']
      raise 'invalid logfile setting: not writable' unless !config['logfile'] or File.writable? config['logfile'] or File.writable? File.dirname(config['logfile'])
      
      # Check whitelist
      raise 'invalid whitelist: set to a file path, or false to disable' unless config['whitelist'] == false or config['whitelist'].is_a? String
      raise "invalid whitelist: cannot open #{config['whitelist']}" unless config['whitelist'] == false or File.readable?(config['whitelist']) 

      # Check external_command
      raise 'invalid external_command: set to a file path, or false to disable' unless config['external_command'].is_a? String or config['external_command'] == false
      raise 'invalid external_command: not executable' unless !config['external_command'] or File.executable?(config['external_command'].split[0])

      # Check no_kill
      raise 'invalid no_kill: set to true or false' unless config['no_kill'].is_bool?

      # Check no_kill_ppid
      raise 'invalid no_kill_ppid: set to true or false' unless config['no_kill_ppid'].is_bool?

      # Check ignore_root_procs
      raise 'invalid ignore_root_procs: set to true or false' unless config['ignore_root_procs'].is_bool?

      # Check log_whitelist
      raise 'invalid log_whitelist: set to true or false' unless config['log_whitelist'].is_bool?

      # Check require_init_wlist
      raise 'invalid require_init_wlist: set to true or false' unless config['require_init_wlist'].is_bool?

      # Check proc_scan_offset
      raise 'invalid proc_scan_offset: set to a number' unless config['proc_scan_offset'].is_a? Fixnum
      raise "invalid proc_scan_offset: set lower than #{@max_pids}" unless config['proc_scan_offset'] < @max_pids

      # Check pidfile
      raise 'invalid pidfile setting: set to a string' unless config['pidfile'].is_a? String
      raise 'invalid pidfile setting: empty' if config['pidfile'].empty?
      raise 'invalid pidfile setting: not writable' unless File.writable? config['pidfile'] or File.writable? File.dirname(config['pidfile'])

      # Return the configuration if no errors encountered
      config
    end

    # Exit cleanly by logging our departure and deleting the pid file if we received a signal
    def clean_exit(sig)
      @log.info "Got SIG#{sig}, exiting."
      File.unlink(@config['pidfile']) if @config['daemon'] and @config['pidfile']
      exit
    end

    # Fetch the current running processes and put them into a ruby set
    # TODO: make this multi platform
    def get_process_list
      # Initialize the set
      entries = Set.new
      # Look up all pids
      Dir.glob('/proc/[0-9]*') { |x|
        # Turn pids to numbers
        pid = File.basename(x).to_i
        # Formal check for unexpected files on /proc starting with a digit
        next unless pid.to_s =~ /^[0-9]+$/
        # Skip processes above the configured offset
        if pid > @config['proc_scan_offset']
          entries.add pid
        end
      }

      return entries
    end

    # Load the whitelist
    def load_whitelist(file)
      # Open the file and read all words from lines that don't start with a #
      begin
        @whitelist = File.read(file).lines.grep(/^[^#]/).join.split
      rescue => e
        puts "Cannot load whitelist: #{e.message}"
        exit
      end
      # Warn the user of the whitelist was loaded but found empty
      if @whitelist.empty? 
        puts 'WARNING: Empty whitelist loaded.'
      end
    end

    # Run the external command
    def run_command(process,parent)

      # Log our action
      @log.info "Running external command: #{@config['external_command']}."

      # Prepare the environment with some useful informations
      ENV['EVIL_PID'] = process.pid.to_s
      ENV['EVIL_CMD'] = process.cmdline
      ENV['EVIL_NAME'] = process.name
      ENV['EVIL_BIN'] = process.binary

      ENV['EVIL_PPID'] = parent.pid.to_s
      ENV['EVIL_PCMD'] = parent.cmdline
      ENV['EVIL_PNAME'] = parent.name
      ENV['EVIL_PBIN'] = parent.binary

      # Launch the command
      system(@config['external_command'])
    end

    # Daemon startup procedure
    def initialize(config)
      # Check max pids first
      if RUBY_PLATFORM.downcase =~ /linux/
        begin
          @max_pids = File.read('/proc/sys/kernel/pid_max').to_i
        rescue
          raise 'Cannot read /proc/sys/kernel/pid_max (is /proc mounted?)'
        end
      else
        # Default on newer FreeBSD. Solaris should have a cap of 30000.
        # Makes only sense once get_process_list becomes multi-platform, but nice to have.
        @max_pids = 99999
      end

      # Validate configuration and store if valid
      @config = validate_conf config

      # Save the full path to the pid file in @config
      @config['pidfile'] = File.expand_path(@config['pidfile'])

      # Get initial process list at startup
      @initial_procs = get_process_list
      raise 'No processes found (is /proc mounted?)' if @initial_procs.empty?

      # Load the whitelist
      load_whitelist @config['whitelist'] if @config['whitelist']

      # Initialize @log as configured
      if @config['syslog']
        @log = Logger::Syslog.new('kunoichi', Syslog::LOG_DAEMON)
      else
        if @config['daemon'] 
          @log = Logger.new(@config['logfile'])
        else
          @log = Logger.new(STDOUT)
        end
      end

      # Enable debug info if running in debug mode, keep it quiet otherwise
      if @config['debug']
        @log.level = Logger::DEBUG
      else
        @log.level = Logger::INFO
      end

      @log.info 'Kunoichi starting up'

      # Daemonize if we're configured for it
      if @config['daemon']
        Process.daemon
        # Only write the pid file if we're running as daemon
        if @config['pidfile']
          begin
            File.open(@config['pidfile'], 'w+') { |x| x.write Process.pid.to_s }
          rescue => e
            @log.error e.message
          end
        end
      end

      # Catch SIGINT (^C) and SIGTERM (default kill)
      [ 'INT', 'TERM' ].each do |signal|
        Signal.trap signal do clean_exit signal end
      end

      # Finally, enter the main loop
      loop do 
        main_loop
        sleep @config['interval']
      end
    end

    # Main loop
    def main_loop
      @log.debug 'Checking processes:'
      # Get running processes
      procs = get_process_list
      # Extract the new processes since last loop cycle
      new_procs = procs - @initial_procs

      # Skip the run if there are new processes
      (@log.debug "\tNo new processes found."; return) if new_procs.empty?

      @log.debug "\tNew processes:"

      # For each new process:
      new_procs.each do |pid|
        # Extract information
        begin
          process = ProcEntry.new(pid)
        rescue
          # Skip if the process died before we could extract the info
          @log.debug "\t\t#{pid} - Disappeared before analysis"
          next
        end
        @log.debug "\t\t#{pid} - #{process.cmdline}"
         
        # If the process runs as root
        if process.uid == 0 

          # Attempt to identify the parent
          begin
            parent = ProcEntry.new(process.parent)
          rescue => e
            @log.warn "\t\tParent process of #{process.name} disappeared (#{e.message})."
            next
          end

          # Skip if the parent has the whitelisted gid
          next if parent.gid == @config['group']

          # If the parent is root
          if parent.uid == 0 
            # ..and we've chosen to ignore root spawned processes
            if @config['ignore_root_procs'] or parent.gid == @config['group']
              # if the parent is init
              if parent.pid == 1
                # spare the process unless configured otherwise
                unless @config['require_init_wlist']
                  next
                end
              # if the parent is not init everything is normal, just skip this one
              else
                next
              end
            end
          else
            # Also skip if the parent is not root, but is in the magic gid
            next if parent.gid == @config['group']
          end

          # Log our finding
          @log.info "\t\tFound offending process: #{pid} - #{process.cmdline}"

          # Skip if the executable is whitelisted
          if @config['whitelist'] and @whitelist.include? process.binary
            @log.info "\t\t\tAllowed(#{process.binary})." if @config['log_whitelist']
            next
          end
          
          # If we got this far it means the process deserves our attention

          # Terminate the process unless configured otherwise
          unless @config['no_kill'] 
            @log.info "\t\t\tTerminating."
            Process.kill(:KILL,pid)
          else
            @log.info "\t\t\tNot killing."
          end

          # Also kill the parent unless configured otherwise (or the parent is init)
          unless process.parent == 1 or @config['no_kill_ppid'] 
            @log.info "\t\t\tKilling parent too."
            Process.kill(:KILL,process.parent) 
          end

          # Run the external command in a subprocess
          fork { run_command(process,parent) } if @config['external_command'] 
        end
      end

      # Save the current process list for the next run
      @initial_procs = procs
    end
  end
end
